Test-isolation leak: tests/run-all (L3/L4) creates empty 'init' commits on the caller's git branch

## Symptom
Running `tests/run-all.sh` in a SWE worktree leaves **empty `git commit` ("init") commits on the worktree's branch**. Observed on two separate tasks this session (#105 readme branches): 3 stray empties on feat/readmes-docs-tests, 2 on feat/readmes-code-surfaces-final, each on top of the real work commit. `git show --stat` confirms they're empty (no file changes). Squash-merge neutralizes them, so no bad content shipped — but any non-squash flow, the verification gate's clean-tree assumptions, or `git log` hygiene are polluted, and it's a latent footgun.

## Cause (to confirm)
A test in L3 (mcp/ or hooks/ integration) or L4 (workflow-sim) sets up a sandbox git repo with `git init` + `git commit` but runs the commit with cwd = the caller's worktree (or without `git -C <sandbox>` / isolated `GIT_DIR`/`GIT_WORK_TREE`), so the commit lands on the real branch instead of the sandbox. The empty + "init" message points at a fixture-repo bootstrap.

## Fix
- Find the test(s) issuing `git commit` and isolate them: operate via `git -C "$SANDBOX"` (or a subshell `cd`), and/or set `GIT_DIR`/`GIT_WORK_TREE` to the sandbox, and use a temp dir outside any tracked repo. Add a guard so a test commit can never target the repo root.
- Optional safety net: a post-run-all check (or the verification gate) that fails if the run created commits on the current branch.

## Related (same worktree-hygiene area)
SWE worktrees ship WITHOUT `mcp/trajectory-server/node_modules` (gitignored), so `tests/run-all` (and the verification gate) fail L3/L4 with `Cannot find module '@modelcontextprotocol/sdk'` until `bun install` is run in `mcp/trajectory-server/`. The gate run is ~208s vs a ~240s budget — cold worktrees are timing-tight. Consider auto-installing deps (or a faster gate path) for worktree runs.

## Acceptance
- `tests/run-all.sh` in a worktree creates ZERO commits on the current branch.
- (related) worktree test runs resolve deps without manual `bun install`, or the gap is documented + the gate budget adjusted.

## Milestone
v0.10.0

--- a/tests/l3-integration/hooks/swe-atomic-close-agent-runs.test.sh
+++ b/tests/l3-integration/hooks/swe-atomic-close-agent-runs.test.sh
@@ -20,15 +20,22 @@ HOOK="$PLUGIN_ROOT/scripts/hooks/swe-atomic-close.sh"
 TMPDIR=$(mktemp -d)
 trap 'rm -rf "$TMPDIR"' EXIT
 
+# Step out of the plugin checkout before any git work so a stray bare git call
+# can never touch the caller's branch, then assert the move stuck. Every commit
+# below targets its sandbox via `git -C "$REPO"`; this is belt-and-suspenders.
+cd "$TMPDIR"
+assert_not_in_plugin_repo "$PLUGIN_ROOT"
+
 # ---- Fixture: repo + two attached worktrees for two same-batch tasks -----
+# Every git write targets the sandbox via `git -C "$REPO"` so cwd never leaks
+# the bootstrap commit onto the caller's branch.
 REPO="$TMPDIR/repo"
 git init -q -b dev "$REPO"
-cd "$REPO"
-git config user.email t@t.io
-git config user.name t
-echo init > README.md
-git add .
-git commit -qm init
+git -C "$REPO" config user.email t@t.io
+git -C "$REPO" config user.name t
+echo init > "$REPO/README.md"
+git -C "$REPO" add .
+git -C "$REPO" commit -qm init
 
 DB="$REPO/.claude/tmb/trajectory.db"
 mkdir -p "$(dirname "$DB")"

--- a/tests/l3-integration/hooks/swe-atomic-close.test.sh
+++ b/tests/l3-integration/hooks/swe-atomic-close.test.sh
@@ -15,21 +15,28 @@ HOOK="$PLUGIN_ROOT/scripts/hooks/swe-atomic-close.sh"
 TMPDIR=$(mktemp -d)
 trap 'rm -rf "$TMPDIR"' EXIT
 
+# Step out of the plugin checkout before any git work so a stray bare git call
+# can never touch the caller's branch, then assert the move stuck. Every commit
+# below targets its sandbox via `git -C "$REPO"`; this is belt-and-suspenders.
+cd "$TMPDIR"
+assert_not_in_plugin_repo "$PLUGIN_ROOT"
+
 # ---- Fixture: bare git remote (plays the role of origin) -----------------
 REMOTE="$TMPDIR/remote.git"
 git init -q --bare "$REMOTE"
 
 # ---- Fixture: main repo (bro's checkout) + detached-HEAD worktree --------
+# Every git write targets the sandbox via `git -C "$REPO"` so cwd never leaks
+# the bootstrap commit onto the caller's branch.
 REPO="$TMPDIR/repo"
 git init -q -b dev "$REPO"
-cd "$REPO"
-git config user.email t@t.io
-git config user.name t
-git remote add origin "$REMOTE"
-echo init > README.md
-git add .
-git commit -qm init
-git push -q origin dev
+git -C "$REPO" config user.email t@t.io
+git -C "$REPO" config user.name t
+git -C "$REPO" remote add origin "$REMOTE"
+echo init > "$REPO/README.md"
+git -C "$REPO" add .
+git -C "$REPO" commit -qm init
+git -C "$REPO" push -q origin dev
 
 DB="$REPO/.claude/tmb/trajectory.db"
 mkdir -p "$(dirname "$DB")"
@@ -85,13 +92,13 @@ export TRAJECTORY_DB_PATH="$DB"
 # the branch ref so SWE's commits advance it directly.
 # Slug = everything after the last '/' in branch_id.
 WT_PATH="$REPO/.claude/worktrees/test-branch"
-git branch fix/test-branch HEAD
-git worktree add -q "$WT_PATH" fix/test-branch
+git -C "$REPO" branch fix/test-branch HEAD
+git -C "$REPO" worktree add -q "$WT_PATH" fix/test-branch
 
 # Worktree for task 44 (needs_validation, slug = nv-branch).
 WT_NV_PATH="$REPO/.claude/worktrees/nv-branch"
-git branch fix/nv-branch HEAD
-git worktree add -q "$WT_NV_PATH" fix/nv-branch
+git -C "$REPO" branch fix/nv-branch HEAD
+git -C "$REPO" worktree add -q "$WT_NV_PATH" fix/nv-branch
 
 swe_input() {
   jq -n '{subagent_type: "swe"}'

--- a/tests/lib/assert.sh
+++ b/tests/lib/assert.sh
@@ -67,6 +67,16 @@ assert_not_contains() {
   fi
 }
 
+assert_not_in_plugin_repo() {
+  local plugin_root="$1"
+  local toplevel
+  toplevel="$(git rev-parse --show-toplevel 2>/dev/null || true)"
+  if [ -n "$toplevel" ] && [ "$toplevel" = "$plugin_root" ]; then
+    printf "${RED}ABORT${NC} test cwd is the plugin repo root (%s) — sandbox isolation breach; refusing to run to avoid committing on the caller branch\n" "$toplevel" >&2
+    exit 1
+  fi
+}
+
 assert_exit_code() {
   local expected="$1"
   local actual="$2"

